### ISSUE TITLE
Make JULIA_NUM_THREADS=auto the default

### ISSUE THREAD

**Issue by jtrakk** [JuliaLang/julia#43672]
This will help new Julia programmers who don't know about `JULIA_NUM_THREADS` get fast execution. It also symbolizes how Julia supports the increasing parallelism of modern technical computing. Anybody who needs sequential execution, e.g. for debugging, can use `JULIA_NUM_THREADS=1`.

- [Rust's Rayon](https://github.com/rayon-rs/rayon/blob/master/FAQ.md#how-many-threads-will-rayon-spawn) uses the number of logical cores by default.

- [Intel® Cilk™ Plus](http://wwweic.eri.u-tokyo.ac.jp/computer/manual/eic2015/doc/intel/en_US/compiler_c/GUID-85B8EB6D-A49C-4D44-B58E-88DCA85FBE23.htm) uses all available hardware threads, which is probably too many for us.

- [Crystal](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html) defaults to 4.

- [Fortran OMP](https://docs.oracle.com/cd/E19205-01/819-5262/aeuip/index.html) defaults to 1.

**Comment by ViralBShah** [comment]
At the very least need a good way to get the physical core count as discussed in #43692

**Comment by fonsp** [comment]
We could use `min(1, Sys.CPU_THREADS ÷ 2)` as a safe default while waiting on that issue? I also think giving more than one thread by default would be great for new users!

**Comment by gbaraldi** [comment]
There is also the issue that new cpus have big.Little architectures (M1 mac, 12th gen intel) so we would have to choose either just performance cores (probably the best) or all cores.

**Comment by lassepe** [comment]
It seems like this issue got stuck. Is there any fundamental downside of having `JULIA_NUM_THREADS=auto,auto` be the default over the current `1,0`? I understand that the auto-detection of the CPU core count is not perfect right now but setting to `1,0` seems to provide a worse default behavior (especially for new users) almost every time (since `Threads.@spawn` and friends will not warn about having only a single thread so it is just silently refusing the multi-thread.)

**Comment by IanButterworth** [comment]
I'm tentatively adding this to the 1.12 milestone, as it seems the direction to go in and would help REPL issues like https://github.com/JuliaLang/julia/issues/55434

My understanding is the main issue is Distributed not being threadsafe? But that may have improved?

**Comment by ViralBShah** [comment]
I do not believe Distributed has seen any major contributions in a long time.

**Comment by jpsamaroo** [comment]
DistributedNext.jl has seen thread-safety changes merged, so users have options if we wanted to switch this on.

That said, if we're going to block this on Distributed.jl becoming thread-safe, then we should have someone from the core team help review https://github.com/JuliaLang/Distributed.jl/pull/101. The changes are not really too complicated, especially if you just review everything before https://github.com/JuliaLang/Distributed.jl/pull/101/commits/0377cda9b4bfa49205305ca7a769d0d6bd75f8f8 (since this commit and those following it are not necessary to improve the thread-safety of Distributed).

**Comment by ViralBShah** [comment]
Should we link to DistributedNext.jl in the Distributed.jl repo?

**Comment by jpsamaroo** [comment]
That would be great! We're registering 1.0 now: https://github.com/JuliaRegistries/General/pull/120499. (We picked 1.0, rather than a pre-release version, since it's just Distributed.jl with fixes on top right now).

**Comment by DilumAluthge** [comment]
Should `import Distributed` print a warning if the number of threads is > 1?

**Comment by jpsamaroo** [comment]
It should probably print a warning on the first use of `remotecall*` et. al - we don't want to spam users who don't actually end up using multiple workers, when Distributed is a deep dependency.

**Comment by ViralBShah** [comment]
Is the idea that DistributedNext.jl will eventually be renamed to Distributed.jl, or will it just be a new independent package?

**Comment by jpsamaroo** [comment]
It's unclear right now what the end result will be, but the plan for now is to keep compatibility with Distributed.jl's API as much as possible to allow for potentially merging back in.

**Comment by IanButterworth** [comment]
Indeed. I think the question remains about whether to start with `-tauto,1` so I'll rename the issue.
